File: data/compilers_comparison/src/examples/second_example/src/func12.c
#include "second_example.h" 
array_t* func12(array_t_param* vars, const unsigned long PATH0, int loopsFactor) {
   size_t pCounter = vars->size;
   if(PATH0 & 1) {
      array_t* array129;
      if (pCounter > 0) {
         array129 = vars->data[--pCounter];
         array129->refC++;
         DEBUG_COPY(array129->id);
      } else {
         array129 = (array_t*)malloc(sizeof(array_t));
         array129->size = 46;
         array129->refC = 1;
         array129->id = 129;
         array129->data = (unsigned int*)malloc(array129->size*sizeof(unsigned int));
         memset(array129->data, 0, array129->size*sizeof(unsigned int));
         DEBUG_NEW(array129->id);
      }
      unsigned int loop39 = 0;
      unsigned int loopLimit39 = (rand()%loopsFactor)/3 + 1;
      for(; loop39 < loopLimit39; loop39++) {
         for (int i = 0; i < array129->size; i++) {
            array129->data[i]--;
         }
         for (int i = 0; i < array129->size; i++) {
            if (array129->data[i] == 49) { 
               return array129;
            }
         }
      }
      array_t_param params0;
      params0.size = 1;
      params0.data = (array_t**)malloc(params0.size*sizeof(array_t*));
      params0.data[0] = array129;
      array_t* array130 = func24(&params0, rng(), loopsFactor);
      DEBUG_RETURN(array130->id);
      free(params0.data);
      array130->refC--;
      if(array130->refC == 0) {
         free(array130->data);
         free(array130);
         DEBUG_FREE(array130->id);
      }
      array129->refC--;
      if(array129->refC == 0) {
         free(array129->data);
         free(array129);
         DEBUG_FREE(array129->id);
      }
   }
   else {
      array_t_param params0;
      params0.size = 0;
      params0.data = (array_t**)malloc(params0.size*sizeof(array_t*));
      array_t* array154 = func16(&params0, rng(), loopsFactor);
      DEBUG_RETURN(array154->id);
      free(params0.data);
      array154->refC--;
      if(array154->refC == 0) {
         free(array154->data);
         free(array154);
         DEBUG_FREE(array154->id);
      }
   }
   array_t* array164;
   if (pCounter > 0) {
      array164 = vars->data[--pCounter];
      array164->refC++;
      DEBUG_COPY(array164->id);
   } else {
      array164 = (array_t*)malloc(sizeof(array_t));
      array164->size = 726;
      array164->refC = 1;
      array164->id = 164;
      array164->data = (unsigned int*)malloc(array164->size*sizeof(unsigned int));
      memset(array164->data, 0, array164->size*sizeof(unsigned int));
      DEBUG_NEW(array164->id);
   }
   return array164;
}

File: data/multilang_comparison/program_nim/src/func12.nim
proc func12*(vars: ptr ArrayTParam, PATH0: uint64, loopsFactor: cint): ptr ArrayT =
   var pCounter = vars.size
   if (PATH0 and 1'u64) != 0:
      discard
      var array161: ptr ArrayT
      if pCounter > 0:
         dec pCounter
         array161 = vars.data[pCounter]
         inc array161.refC
         debugCopy(array161.id)
      else:
         array161 = cast[ptr ArrayT](alloc(sizeof(ArrayT)))
         array161.size = 975
         array161.refC = 1
         array161.id = 161
         array161.data = cast[ptr UncheckedArray[uint32]](alloc(array161.size * sizeof(uint32)))
         zeroMem(array161.data, array161.size * sizeof(uint32))
         debugNew(array161.id)
      let loopLimit48 = (50) div (3) + 1
      for loop48 in 0 ..< loopLimit48:
         discard
         for i in 0 ..< array161.size:
            dec array161.data[i]
         for i in 0 ..< array161.size:
            if array161.data[i] == 29:
               return array161
      var params0: ArrayTParam
      params0.size = 1
      params0.data = cast[ptr UncheckedArray[ptr ArrayT]](alloc(params0.size * sizeof(ptr ArrayT)))
      params0.data[0] = array161
      var array162: ptr ArrayT = func24(addr params0, getPath(), loopsFactor)
      debugReturn(array162.id)
      dealloc(params0.data)
      dec array162.refC
      if array162.refC == 0:
         dealloc(array162.data)
         dealloc(array162)
         debugFree(array162.id)
      dec array161.refC
      if array161.refC == 0:
         dealloc(array161.data)
         dealloc(array161)
         debugFree(array161.id)
   else:
      discard
      var params0: ArrayTParam
      params0.size = 0
      params0.data = cast[ptr UncheckedArray[ptr ArrayT]](alloc(params0.size * sizeof(ptr ArrayT)))
      var array186: ptr ArrayT = func16(addr params0, getPath(), loopsFactor)
      debugReturn(array186.id)
      dealloc(params0.data)
      dec array186.refC
      if array186.refC == 0:
         dealloc(array186.data)
         dealloc(array186)
         debugFree(array186.id)
   var array196: ptr ArrayT
   if pCounter > 0:
      dec pCounter
      array196 = vars.data[pCounter]
      inc array196.refC
      debugCopy(array196.id)
   else:
      array196 = cast[ptr ArrayT](alloc(sizeof(ArrayT)))
      array196.size = 552
      array196.refC = 1
      array196.id = 196
      array196.data = cast[ptr UncheckedArray[uint32]](alloc(array196.size * sizeof(uint32)))
      zeroMem(array196.data, array196.size * sizeof(uint32))
      debugNew(array196.id)
   return array196

